pkg/apis/pipeline/v1alpha1/pipeline_validation_test.go

```diff
@@ -214,244 +213,616 @@ func TestPipeline_Validate(t *testing.T) {
+			Spec: v1alpha1.PipelineSpec{
+				Params: []v1alpha1.ParamSpec{{
+					Name:    "baz",
+					Type:    v1alpha1.ParamTypeString,
+					Default: v1beta1.NewArrayOrString("an", "array"),
+				}},
+				Tasks: []v1alpha1.PipelineTask{{
+					Name:    "foo",
+					TaskRef: &v1alpha1.TaskRef{Name: "foo-task"},
+				}},
+			},
+		},
 		failureExpected: true,
 	}, {
 		name: "array parameter used as string",
-		p: tb.Pipeline("pipeline", tb.PipelineSpec(
-			tb.PipelineParamSpec("baz", v1alpha1.ParamTypeArray, tb.ParamSpecDefault("anarray", "elements")),
-			tb.PipelineTask("bar", "bar-task",
-				tb.PipelineTaskParam("a-param", "$(params.baz)")),
-		)),
+		p: &v1alpha1.Pipeline{
+			ObjectMeta: metav1.ObjectMeta{Name: "pipeline"},
+			Spec: v1alpha1.PipelineSpec{
+				Params: []v1alpha1.ParamSpec{{
+					Name:    "baz",
+					Type:    v1alpha1.ParamTypeArray,
+					Default: v1beta1.NewArrayOrString("an", "array"),
+				}},
+				Tasks: []v1alpha1.PipelineTask{{
+					Name:    "foo",
+					TaskRef: &v1alpha1.TaskRef{Name: "foo-task"},
+					Params: []v1alpha1.Param{{
+						Name:  "a-param",
+						Value: *v1beta1.NewArrayOrString("$(params.baz)"),
+					}},
+				}},
+			},
+		},
 		failureExpected: true,
 	}, {
 		name: "star array parameter used as string",
-		p: tb.Pipeline("pipeline", tb.PipelineSpec(
-			tb.PipelineParamSpec("baz", v1alpha1.ParamTypeArray, tb.ParamSpecDefault("anarray", "elements")),
-			tb.PipelineTask("bar", "bar-task",
-				tb.PipelineTaskParam("a-param", "$(params.baz[*])")),
-		)),
+		p: &v1alpha1.Pipeline{
+			ObjectMeta: metav1.ObjectMeta{Name: "pipeline"},
+			Spec: v1alpha1.PipelineSpec{
+				Params: []v1alpha1.ParamSpec{{
+					Name:    "baz",
+					Type:    v1alpha1.ParamTypeArray,
+					Default: v1beta1.NewArrayOrString("an", "array"),
+				}},
+				Tasks: []v1alpha1.PipelineTask{{
+					Name:    "foo",
+					TaskRef: &v1alpha1.TaskRef{Name: "foo-task"},
+					Params: []v1alpha1.Param{{
+						Name:  "a-param",
+						Value: *v1beta1.NewArrayOrString("$(params.baz[*])"),
+					}},
+				}},
+			},
+		},
 		failureExpected: true,
 	}, {
 		name: "array parameter string template not isolated",
-		p: tb.Pipeline("pipeline", tb.PipelineSpec(
-			tb.PipelineParamSpec("baz", v1alpha1.ParamTypeArray, tb.ParamSpecDefault("anarray", "elements")),
-			tb.PipelineTask("bar", "bar-task",
-				tb.PipelineTaskParam("a-param", "first", "value: $(params.baz)", "last")),
-		)),
+		p: &v1alpha1.Pipeline{
+			ObjectMeta: metav1.ObjectMeta{Name: "pipeline"},
+			Spec: v1alpha1.PipelineSpec{
+				Params: []v1alpha1.ParamSpec{{
+					Name:    "baz",
+					Type:    v1alpha1.ParamTypeArray,
+					Default: v1beta1.NewArrayOrString("an", "array"),
+				}},
+				Tasks: []v1alpha1.PipelineTask{{
+					Name:    "foo",
+					TaskRef: &v1alpha1.TaskRef{Name: "foo-task"},
+					Params: []v1alpha1.Param{{
+						Name:  "a-param",
+						Value: *v1beta1.NewArrayOrString("first", "value: $(params.baz)", "last"),
+					}},
```

Comment:
Nothing related to changes in this PR, but wouldnt this test case fail with `not defined param`  just like one [above](https://github.com/tektoncd/pipeline/pull/3250/files#diff-0f59e70a1c4fe64fff61b4105c80e00aR603)?

Reply:
Yeah, good catch. I fixed it, but this highlights the fact that these tests should probably compare the actual returned error, not just whether an error was returned. Simple typos can easily cause cases to go untested this way 😬

Reply:
yup absolutely 👍